carsus/io/output/base.py

```diff
@@ -0,0 +1,546 @@
+                                     > levels_metastable_loggf_threshold]
+
+        # Count the remaining strong transitions
+        metastable_lines_grouped = metastable_lines.groupby("upper_level_id")
+        metastable_counts = metastable_lines_grouped["upper_level_id"].count()
+        metastable_counts.name = "metastable_counts"
+
+        # If there are no strong transitions for a level (the count is NaN)
+        # then the metastable flag is True else (the count is a natural number)
+        # the metastable flag is False
+        levels = levels.join(metastable_counts)
+        metastable_flags = levels["metastable_counts"].isnull()
+        metastable_flags.name = "metastable"
+
+        return metastable_flags
+
+    def _get_all_levels_data(self):
+        """ Returns the same output than `AtomData._get_all_levels_data()` """
+        gf = self.gfall_reader
+        gf.levels['g'] = 2*gf.levels['j'] + 1
+        gf.levels['g'] = gf.levels['g'].map(np.int)
+
+        ions_df = pd.DataFrame.from_records(
+            self.ions, columns=["atomic_number", "ion_charge"])
+        ions_df = ions_df.set_index(['atomic_number', 'ion_charge'])
+        levels = gf.levels.reset_index().join(ions_df, how="inner",
+                                              on=["atomic_number",
+                                                  "ion_charge"]).\
+            set_index(["atomic_number", "ion_charge"])
+        levels = levels.drop(columns=['j', 'label', 'method'])
+        levels['level_id'] = range(1, len(levels)+1)
+        levels = levels.reset_index().reset_index(drop=True)
+        levels = levels.rename(columns={'ion_charge': 'ion_number'})
+        levels = levels[['atomic_number', 'ion_number', 'g', 'energy']]
+
+        levels['energy'] = levels['energy'].apply(lambda x: x*u.Unit('cm-1'))
+        levels['energy'] = levels['energy'].apply(
+            lambda x: x.to(u.eV, equivalencies=u.spectral()))
+        levels['energy'] = levels['energy'].apply(lambda x: x.value)
+
+        ground_levels = self.ground_levels
+        ground_levels.rename(
+            columns={'ion_charge': 'ion_number'}, inplace=True)
+
+        # Fixes Ar II duplicated ground level. For Kurucz, ground state
+        # has g=2, for NIST has g=4. We keep Kurucz.
+        mask = (ground_levels['atomic_number'] == 18) & (
+            ground_levels['ion_number'] == 1)
+        ground_levels.loc[mask, 'g'] = 2
+
+        levels = pd.concat([ground_levels, levels], sort=True)
+        levels['level_id'] = range(1, len(levels)+1)
+        levels = levels.set_index('level_id')
+        levels = levels.drop_duplicates(keep='last')
+
+        return levels
+
+    def _get_all_lines_data(self, levels):
+        """ Returns the same output than `AtomData._get_all_lines_data()` """
+        gf = self.gfall_reader
+        df_list = []
+
+        for ion in self.ions:
+
+            try:
+                df = gf.lines.loc[ion]
+
+            except (KeyError, TypeError):
```

Comment:
Do we want to ignore `TypeErrors`?

Reply:
Yes. It's strange, but when using MultiIndex the `.loc` method sometimes throws `KeyError` and sometimes `TypeError`. This is a popular question in the Pandas issue page and the developers say it's not a bug.